core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceServiceSpec.scala

```
@@ -404,135 +393,186 @@ class MultiCloudWorkspaceServiceSpec extends AnyFlatSpec with Matchers with Opti
 
-    doReturn(Future.successful(testData.azureWorkspace))
-      .when(mcWorkspaceService)
-      .getWorkspaceContext(equalTo(testData.azureWorkspace.toWorkspaceName), any())
+    runTest(mcWorkspaceService)
+  }
 
-    val result = intercept[RawlsExceptionWithErrorReport] {
-      Await.result(
-        mcWorkspaceService.cloneMultiCloudWorkspace(
-          mock[WorkspaceService](RETURNS_SMART_NULLS),
-          testData.azureWorkspace.toWorkspaceName,
-          WorkspaceRequest(
-            testData.billingProject.projectName.value,
-            UUID.randomUUID().toString,
-            Map.empty,
-            None,
-            None,
-            None,
-            None
+  it should "fail if the destination workspace already exists" in
+    withEmptyTestDatabase {
+      withMockedMultiCloudWorkspaceService { mcWorkspaceService =>
+        val result = intercept[RawlsExceptionWithErrorReport] {
+          Await.result(
+            for {
+              _ <- slickDataSource.inTransaction { access =>
+                access.rawlsBillingProjectQuery.create(testData.billingProject) >>
+                  access.workspaceQuery.createOrUpdate(testData.azureWorkspace)
+              }
+              _ <- mcWorkspaceService.cloneMultiCloudWorkspace(
+                mock[WorkspaceService](RETURNS_SMART_NULLS),
+                testData.azureWorkspace.toWorkspaceName,
+                WorkspaceRequest(
+                  testData.azureWorkspace.namespace,
+                  testData.azureWorkspace.name,
+                  Map.empty
+                )
+              )
+            } yield fail(),
+            Duration.Inf
           )
-        ),
-        Duration.Inf
-      )
+        }
+
+        result.errorReport.statusCode.value shouldBe StatusCodes.Conflict
+      }
     }
 
-    result.errorReport.statusCode.value shouldBe StatusCodes.BadRequest
-  }
+  it should "not create a workspace record if the request to Workspace Manager fails" in
```

Will also want to test for an error being in the return result object.

---

I'm going to let the runner handle that

---

But you can't let the runner handle it if there is no returned job ID. And even if there is a returned Job id with an error status, I'm not sure it makes sense to have the runner poll again and do the cleanup (though I guess that helps isolate cleanup code?).

---

> if there is no returned job ID

So you're telling me that it doesn't always return an ID?

> though I guess that helps isolate cleanup code?

That's exactly what I'm going after

---

I am still thinking we need coverage for the case of no job ID.

---

Are there cases when WSM returns a non-error code but does not return a job id? If there are [strong opinion], that's weird and should be fixed.

---

I don't think it's weird. A job ID will only be created if there is a job to be executed. If an error is encountered that prevents getting to the point of creating the job, I would expect an ErrorReport and no JobReport. But we can discuss at standup to get the opinions of others.

---

Also might be good to ask in the terra-workspace-manager Slack channel about what the expectations are. Should one always assume a JobReport will be returned (with a jobId), or should one handle the case of no JobReport and inspect the ErrorReport).

---

Ah, I missed non-error code in your message. Yes, I would expect an error code in that case. Are you then logging the message in the ErrorReport if the server call has a non-success status?

---

We discussed in standup and feel like anything that would prevent a JobReport from being created should return an error status code.